### PR TITLE
SingleR data export to ExpHub

### PR DESCRIPTION
**Pull request by friedue**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,9 @@
+*.Rproj
+.Rproj.user
+.Rhistory
+.RData
+.Ruserdata
+.Rbuildignore
```
Comment (LTLA) [inline .gitignore]: If you add a `.gitignore`, you also need an `.Rbuildignore` to ignore the former during build.

File: R/create_se.R [hunk 2]
```diff
@@ -0,0 +1,26 @@
+#' @importFrom ExperimentHub ExperimentHub
+#' @importFrom SummarizedExperiment SummarizedExperiment
+.create_se <- function(dataset, hub = ExperimentHub(), assays="normcounts",
+                       has.rowdata=TRUE, has.coldata=TRUE, suffix=NULL) {
+    host <- file.path("singleR", dataset)
```
Comment (LTLA) [inline R/create_se.R]: Capital `SingleR`.

File: R/get_data_hpca.R [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+#' 
+#' Processing described in Aran, Looney & Liu et al. (2019). 
+#' \emph{Nature Immunology} 20, 163–172. doi: 10.1038/s41590-018-0276-y
+#' 
+#' @examples
+#' hpca <- refData_hpca()
+#' 
+#' @export
+#' @importFrom SummarizedExperiment rowData
+get_data_hpca <- function() {
```
Comment (LTLA) [inline R/get_data_hpca.R]: Camel case and something like `HPCAData()` is preferred. In fact, I would even go with `HumanPrimaryCellAtlasData()`. Can't be clearer than that. And add an option to prune out NAs.

File: inst/scripts/make-hpca-data.Rmd [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+library(SingleR)
+hpca <- getReferenceDataset(dataset="hpca") ## retrieves and loads RDA from Dvir's repo
+pred.hpca <- SingleR(test = hESCs, training = hpca$data, labels = hpca$types)
+```
+
+
+```{r}
+## retrieving the original data from the legacy github repo
+library(SingleR)
+hpca.ref <- getReferenceDataset(dataset="hpca")
```
Comment (LTLA) [inline inst/scripts/make-hpca-data.Rmd]: Don't use `getReferenceDataset`, because that function itself will be removed once we pull down stuff from EHub. Rather, copy the contents of `getReferenceDataset` into the Rmarkdown document directly.

File: inst/scripts/make-hpca-data.Rmd [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+## get normalized exprs matrix
+normcounts <- hpca.ref$data
+
+## get meta data
+meta <- DataFrame(row.names = colnames(normcounts),
+                       label.main = hpca.ref$main_types,
+                       label.fine = hpca.ref$types)
+
+## construct SE
+se <- SummarizedExperiment(assays = list(normcounts = normcounts),
```
Comment (LTLA) [inline inst/scripts/make-hpca-data.Rmd]: Not really any need to construct `se` at this point.

File: inst/scripts/make-hpca-metadata.R [hunk 6]
```diff
@@ -0,0 +1,27 @@
+write.csv(file="../extdata/metadata-hpca.csv", 
+          data.frame(
+              Title = sprintf("Human Primary Cell Atlas data %s", c("normcounts", "colData")), # This can be the exact file name (if self-describing) or a more complete description.
+              Description = sprintf("%s human primary cell atlas data with 38 main cell types (169 subtypes)", 
+                                    c("Matrix or norm. expression", "Per-sample metadata: cell type labels")),
```
Comment (LTLA) [inline inst/scripts/make-hpca-metadata.R]: The `Description` should be a full english sentence, so the format string should be something like:

- Matrix of normalized expression values from the Human Primary Cell ...
- Per-sample metadata containing cell type labels from the Human ...

File: vignettes/documentation_singleRData.Rmd [hunk 7]
```diff
@@ -0,0 +1,56 @@
+---
```
Comment (LTLA) [inline vignettes/documentation_singleRData.Rmd]: Is this document necessary? Just move the details into the function documentation, with a brief mention about the available datasets in the main vignette (if it isn't there already).

Comment (friedue) [reply vignettes/documentation_singleRData.Rmd]: I'd like to keep this for now as a means to collect whatever tidbits of information I can get; it's still very much a draft, but eventaully I'd like to include more details about the actual cell types present in the individual data sets so that users can judge whether their cell types of interest are better represented by one reference data set over another.

File: inst/scripts/make-blueprint_encode-data.Rmd [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+date: "8/9/2019"
+output: html_document
+---
+
+```{r setup, include=FALSE}
+library(BiocStyle)
+knitr::opts_chunk$set(error=FALSE, message=FALSE, warning=FALSE)
+```
+
+Dvir Aran's original `SingleR` github repository contains Robjects with normalized expression values from reference data sets such as those collected by Blueprint & Encode, the Human Primary Cell Atlas, and ImmGen.
```
Comment (LTLA) [inline inst/scripts/make-blueprint_encode-data.Rmd]: If you use `r Githubpkg("dviraran/SingleR")`, you get a link for free.

File: inst/scripts/make-blueprint_encode-metadata.R [hunk 9]
```diff
@@ -0,0 +1,27 @@
+write.csv(file="../extdata/metadata-blueprint_encode.csv", 
+          data.frame(
+              Title = sprintf("Blueprint/Encode RNA-seq data: %s", c("normcounts", "colData")), # This can be the exact file name (if self-describing) or a more complete description.
```
Comment (LTLA) [inline inst/scripts/make-blueprint_encode-metadata.R]: If you'll forgive the nitpicking:

- I generally like the `Title` to be grammatically simple, so no colons; something like `Blueprint/ENCODE RNA-seq normcounts` and `Blueprint/ENCODE RNA-seq colData`.
- If you're spelling things out anyway, `norm.`  should be `normalized`.

Comment (friedue) [reply inst/scripts/make-blueprint_encode-metadata.R]: lol, you do you

File: inst/scripts/make-blueprint_encode-data.Rmd (hunk 8 above)
Comment (friedue) [reply inst/scripts/make-blueprint_encode-data.Rmd]: nice

**Comment by LTLA** [comment]
Don't forget to merge with `master` regularly, otherwise there'll be pain later.

File: R/dataRetrieval.R [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+#' @examples
+#' hpca.se <- HumanPrimaryCellAtlasData()
+#' 
+#' @export
+#' @importFrom SummarizedExperiment rowData
+HumanPrimaryCellAtlasData <- function() {
+    version <- "1.0.0"
+    se <- .create_se(file.path("hpca", version),
+                     assays="normcounts", rm_NA = NULL,
+                     has.rowdata = FALSE, has_coldata = TRUE,
```
Comment (LTLA) [inline R/dataRetrieval.R]: Where did the underscore come from?

Comment (LTLA) [reply R/dataRetrieval.R]: Need to expose `rm.na=c("row", "column", "both", "none")` in the argument of the user-visible function. And then `match.arg()` before passing the value to `.create_se`.

Comment (friedue) [reply R/dataRetrieval.R]: Exposed it in `BlueprintEncodeData`, not in the other one because it wasn't needed there. Or are you referring to the entire set of options? Will read up on `match.arg`.

Underscores are what dots turn into after 8pm.

Comment (LTLA) [reply R/dataRetrieval.R]: Right, if there aren't any NAs, then it's fine. Probably also want `rm.na` instead of `rm_na`.

Comment (friedue) [reply R/dataRetrieval.R]: More dots it is. I've set `rm.NA = NULL` which should avoid triggering the `is.na()` check in `create_se()`

File: R/dataRetrieval.R [hunk 11]
```diff
@@ -0,0 +1,308 @@
...
+                       has.rowdata=FALSE, has.coldata=TRUE) {
+    
+    ## TEMPORARY CODE for pulling data directly from github until it is 
+    ## available on ExHub =====================================================
+    ref.set <- getReferenceDataset(dataset=dataset)
+    nrmcnts <- ref.set$data
+    colnames(nrmcnts) <- paste(colnames(nrmcnts), 1:ncol(nrmcnts), sep = ".")
+    
+    ## handling NA's
+    if(!is.null(rm.NA)){
```
Comment (LTLA) [inline R/dataRetrieval.R]: Why not allow this to be handled by `"none"`? There's no need for special `NULL` behavior.

File: R/dataRetrieval.R [hunk 12]
```diff
@@ -0,0 +1,308 @@
...
+        keep_rows <- apply(mat, 1, function(x) all(!is.na(x)))
+        if( length(keep_rows[keep_rows]) > 0){
+            mat <- mat[keep_rows, ]
+        }else{
+            warning("The number of rows (genes) with NAs equals the *total* number of genes. We're going to ignore your wish to eliminate rows with NAs.")
+        }
+        
+    }
+    
+    if(rm.NA == "cols" | rm.NA == "both"){
```
Comment (LTLA) [inline R/dataRetrieval.R]: `||` for scalars, `|` for vectors. The second term is not evaluated by `||` if the first term is true.

File: vignettes/SingleR.Rmd [hunk 13]
```diff
@@ -208,6 +208,7 @@ To illustrate, we will use a hESC dataset from `r Biocpkg("scRNAseq")` [@lamanno
 Again, we will subset this test dataset for the sake of speed.
 
 ```{r}
+hESCs <- scRNAseq::LaMannoBrainData('human-es')
```
Comment (LTLA) [inline vignettes/SingleR.Rmd]: Why is this line here?

File: vignettes/SingleR.Rmd [hunk 14]
```diff
@@ -216,9 +217,10 @@ We subset to the common genes and we generate log-normalized expression values a
 
 ```{r}
 common <- intersect(rownames(hESCs), rownames(hpca$data))
-hpca$data <- hpca$data[common,]
 hESCs <- hESCs[common,]
 hESCs <- logNormCounts(hESCs)
+
+hpca$data <- hpca$data[common,]
```
Comment (LTLA) [inline vignettes/SingleR.Rmd]: This should be adjusted to deal with an SE `hpca`.

File: R/dataRetrieval.R (hunk 11 above)
Comment (friedue) [reply R/dataRetrieval.R]: personal indulgence.
